### PR TITLE
bugfix：deadlock when pressure test. ON DUPLICATE KEY UPDATE

### PR DESCRIPTION
**Pull request by huan415**
### Ⅰ. Describe what this PR did
ON DUPLICATE KEY UPDATE.  在压测的时候发生死锁

### CODE REVIEW

**Comment by huan415** [comment]
##以下例子只是证明 for update+on duplicate key update 会发生死锁
public class Test {
    public static void main(String[] args) {
        for (int i = 0; i < 1000; i++) {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    Test t = new Test();
                    t.run();
                }
            }).start();
        }
    }

    private void run() {
        String url = "jdbc:mysql://127.0.0.1:3306/account" ;
        String username = "root" ;
        String password = "root" ;
        Connection connection = null;
        try{

            connection = DriverManager.getConnection(url , username , password ) ;
            connection.setAutoCommit(false);
            String sql0 = "select * from test where f1='1' and f2 = '2'  for update";
            PreparedStatement ps0 = connection.prepareStatement(sql0);
            ps0.executeQuery();

            //error: Deadlock found when trying to get lock; try restarting transaction
            //String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?) ";
            //error: Deadlock found when trying to get lock; try restarting transaction
            String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?)  on duplicate key update f3 =VALUES(f3)";
            //正常
            //String sql = "update test set f1=?,f2=?,f3=? where f1='1' and f2 = '2' ";
            PreparedStatement ps = connection.prepareStatement(sql);
            ps.setString(1, "1");
            ps.setString(2, "2");
            ps.setString(3, "3");
            ps.executeUpdate();
            connection.commit();
        }catch(SQLException se){
            System.out.println("数据库连接失败！");
            se.printStackTrace() ;
            try {
                connection.rollback();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }
}

**Comment by caohdgege** [comment]
> ##以下例子只是证明 for update+on duplicate key update 会发生死锁
> public class Test {
> public static void main(String[] args) {
> for (int i = 0; i < 1000; i++) {
> new Thread(new Runnable() {
> @override
> public void run() {
> Test t = new Test();
> t.run();
> }
> }).start();
> }
> }
> 
> ```
> private void run() {
>     String url = "jdbc:mysql://127.0.0.1:3306/account" ;
>     String username = "root" ;
>     String password = "root" ;
>     Connection connection = null;
>     try{
> 
>         connection = DriverManager.getConnection(url , username , password ) ;
>         connection.setAutoCommit(false);
>         String sql0 = "select * from test where f1='1' and f2 = '2'  for update";
>         PreparedStatement ps0 = connection.prepareStatement(sql0);
>         ps0.executeQuery();
> 
>         //error: Deadlock found when trying to get lock; try restarting transaction
>         //String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?) ";
>         //error: Deadlock found when trying to get lock; try restarting transaction
>         String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?)  on duplicate key update f3 =VALUES(f3)";
>         //正常
>         //String sql = "update test set f1=?,f2=?,f3=? where f1='1' and f2 = '2' ";
>         PreparedStatement ps = connection.prepareStatement(sql);
>         ps.setString(1, "1");
>         ps.setString(2, "2");
>         ps.setString(3, "3");
>         ps.executeUpdate();
>         connection.commit();
>     }catch(SQLException se){
>         System.out.println("数据库连接失败！");
>         se.printStackTrace() ;
>         try {
>             connection.rollback();
>         } catch (SQLException e1) {
>             e1.printStackTrace();
>         }
>     }
> }
> ```
> 
> }

你这个测试demo的建表语法也发一下？

**Comment by caohdgege** [comment]
如果在RR的隔离级别下，这个没有for update可能也没问题，

**Comment by huan415** [comment]
> > ##以下例子只是证明 for update+on duplicate key update 会发生死锁
> > public class Test {
> > public static void main(String[] args) {
> > for (int i = 0; i < 1000; i++) {
> > new Thread(new Runnable() {
> > @override
> > public void run() {
> > Test t = new Test();
> > t.run();
> > }
> > }).start();
> > }
> > }
> > ```
> > private void run() {
> >     String url = "jdbc:mysql://127.0.0.1:3306/account" ;
> >     String username = "root" ;
> >     String password = "root" ;
> >     Connection connection = null;
> >     try{
> > 
> >         connection = DriverManager.getConnection(url , username , password ) ;
> >         connection.setAutoCommit(false);
> >         String sql0 = "select * from test where f1='1' and f2 = '2'  for update";
> >         PreparedStatement ps0 = connection.prepareStatement(sql0);
> >         ps0.executeQuery();
> > 
> >         //error: Deadlock found when trying to get lock; try restarting transaction
> >         //String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?) ";
> >         //error: Deadlock found when trying to get lock; try restarting transaction
> >         String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?)  on duplicate key update f3 =VALUES(f3)";
> >         //正常
> >         //String sql = "update test set f1=?,f2=?,f3=? where f1='1' and f2 = '2' ";
> >         PreparedStatement ps = connection.prepareStatement(sql);
> >         ps.setString(1, "1");
> >         ps.setString(2, "2");
> >         ps.setString(3, "3");
> >         ps.executeUpdate();
> >         connection.commit();
> >     }catch(SQLException se){
> >         System.out.println("数据库连接失败！");
> >         se.printStackTrace() ;
> >         try {
> >             connection.rollback();
> >         } catch (SQLException e1) {
> >             e1.printStackTrace();
> >         }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > }
> 
> 你这个测试demo的建表语法也发一下？



> > ##以下例子只是证明 for update+on duplicate key update 会发生死锁
> > public class Test {
> > public static void main(String[] args) {
> > for (int i = 0; i < 1000; i++) {
> > new Thread(new Runnable() {
> > @override
> > public void run() {
> > Test t = new Test();
> > t.run();
> > }
> > }).start();
> > }
> > }
> > ```
> > private void run() {
> >     String url = "jdbc:mysql://127.0.0.1:3306/account" ;
> >     String username = "root" ;
> >     String password = "root" ;
> >     Connection connection = null;
> >     try{
> > 
> >         connection = DriverManager.getConnection(url , username , password ) ;
> >         connection.setAutoCommit(false);
> >         String sql0 = "select * from test where f1='1' and f2 = '2'  for update";
> >         PreparedStatement ps0 = connection.prepareStatement(sql0);
> >         ps0.executeQuery();
> > 
> >         //error: Deadlock found when trying to get lock; try restarting transaction
> >         //String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?) ";
> >         //error: Deadlock found when trying to get lock; try restarting transaction
> >         String sql = "insert into test (id,f1,f2,f3) values (null,?,?,?)  on duplicate key update f3 =VALUES(f3)";
> >         //正常
> >         //String sql = "update test set f1=?,f2=?,f3=? where f1='1' and f2 = '2' ";
> >         PreparedStatement ps = connection.prepareStatement(sql);
> >         ps.setString(1, "1");
> >         ps.setString(2, "2");
> >         ps.setString(3, "3");
> >         ps.executeUpdate();
> >         connection.commit();
> >     }catch(SQLException se){
> >         System.out.println("数据库连接失败！");
> >         se.printStackTrace() ;
> >         try {
> >             connection.rollback();
> >         } catch (SQLException e1) {
> >             e1.printStackTrace();
> >         }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > }
> 
> 你这个测试demo的建表语法也发一下？

CREATE TABLE `uk`  (
  `id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
  `f1` int(10) UNSIGNED DEFAULT NULL,
  `f2` int(10) UNSIGNED DEFAULT NULL,
  `f3` int(10) UNSIGNED DEFAULT NULL,
  PRIMARY KEY (`id`) USING BTREE,
  UNIQUE INDEX `f1`(`f1`) USING BTREE
) ENGINE = InnoDB AUTO_INCREMENT = 73 CHARACTER SET = utf8 COLLATE = utf8_general_ci ROW_FORMAT = Dynamic;

**Comment by a364176773** [comment]
去了for update你怎么保证前镜像正确？

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/mysql/MySQLInsertOrUpdateExecutor.java [hunk 1]
```diff
@@ -250,7 +250,7 @@ public TableRecords beforeImage() throws SQLException {
      */
     public TableRecords buildTableRecords2(TableMeta tableMeta, String selectSQL, ArrayList<List<Object>> paramAppenderList) throws SQLException {
         ResultSet rs = null;
-        try (PreparedStatement ps = statementProxy.getConnection().prepareStatement(selectSQL + " FOR UPDATE")) {
+        try (PreparedStatement ps = statementProxy.getConnection().prepareStatement(selectSQL)) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/mysql/MySQLInsertOrUpdateExecutor.java:253]: There is no way to ensure data consistency between beforeImage, execution, and afterImage.

**Comment by slievrly** [comment]
@huan415 if you have any improvement，Pls reopen it.

**Pull request closed without merging**